crates/metrics/src/label.rs

```diff
@@ -0,0 +1,545 @@
+
+pub type Labeled<T, A> = WithLabel<A, T>;
+pub type Labeled2<T, A, B> = WithLabel<A, WithLabel<B, T>>;
+pub type Labeled3<T, A, B, C> = WithLabel<A, WithLabel<B, WithLabel<C, T>>>;
+pub type Labeled4<T, A, B, C, D> = WithLabel<A, WithLabel<B, WithLabel<C, WithLabel<D, T>>>>;
+
+pub trait DynamicLabel<M> {
+    type MetricCollection;
+}
+
+pub trait ResolveLabels<L> {
+    type Target;
+
+    fn resolve_labels(&self, labels: L) -> &Self::Target;
+}
+
+/// Adds dynamically resolved (in runtime) label to `M`.
+pub struct WithLabel<L: DynamicLabel<M>, M> {
+    collection: L::MetricCollection,
+}
+
+impl<L, M> WithLabel<L, M>
+where
+    L: DynamicLabel<M>,
+{
+    /// Resolves a single dynamic label and finds the underlying metric.
+    pub fn resolve_label<T>(&self, label: T) -> &M
+    where
+        Self: ResolveLabels<(T,), Target = M>,
+    {
+        self.resolve_labels((label,))
+    }
+
+    /// Resolves multilpe dynamic labels and finds the underlying metric.
+    ///
+    /// Expects a tuple of labels as `LS` argument.
+    pub fn resolve_labels<LS>(&self, labels: LS) -> &<Self as ResolveLabels<LS>>::Target
+    where
+        Self: ResolveLabels<LS>,
+    {
+        ResolveLabels::resolve_labels(self, labels)
+    }
+}
+
+/// Metric label using an `enum` as its value.
+///
+/// The most efficient way to specify metric labels in runtime, metric lookups
+/// using enum labels are as fast as array indexing. Prefer using this when all
+/// of your possible label values are known at the compile time.
+#[derive(Clone, Copy, Debug)]
+pub struct EnumLabel<const NAME: LabelName, T>(T);
+
+impl<const NAME: LabelName, T> EnumLabel<NAME, T> {
+    /// Creates a new [`EnumLabel`].
+    pub fn new(e: impl Into<T>) -> Self {
+        Self(e.into())
+    }
+
+    /// Convert this [`EnumLabel`] into the inner [`Enum`].
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+/// `enum` used as the value in [`EnumLabel`]s.
+///
+/// To implement this `trait` you also need to derive [`Ordinalize`] for your
+/// `enum`.
+/// SAFETY: DO NOT use custom discriminant values (eg. `enum MyEnum { MyVariant
+/// = -1 }`), this will lead to either:
+/// - `panic` in runtime (only for builds with `debug_assertions`)
+/// - incorrect label resolution
+pub trait Enum: Copy + Ordinalize<VariantType = i8> {
+    /// String representation of this enum.
+    fn as_str(&self) -> &'static str;
+}
+
+impl<const NAME: LabelName, T, M> DynamicLabel<M> for EnumLabel<NAME, T>
+where
+    T: Enum,
+{
+    // TODO: Switch to `[(T, M); T::VARIANT_COUNT]` once generic parameters are
+    // allowed to be used in const expressions.
+    type MetricCollection = Vec<(T, M)>;
+}
+
+impl<const NAME: LabelName, T, M> Metric for WithLabel<EnumLabel<NAME, T>, M>
+where
+    T: Enum,
+    M: Metric,
+{
+    fn register(attrs: &Attrs) -> Self {
+        let name = const { resolve_label_name::<NAME>() };
+
+        let metrics = T::VARIANTS.iter().map(|l| {
```

Comment:
Can we please avoid using single character variable names? Makes even harder to review. Please use some meaningful names.

Reply:
It's a closure argument, and the closure is 2 lines long. Usually it's considered ok to have single letter variables if the scope is small.
But I'll change this to `label` just for you :smile:

Reply:
No, I lied, I can't change it to `label` as there's another variable with this name in the scope :clown_face: 

Do you prefer `lbl` +`label` or `label` + `label_` ?

Reply:
Maybe `variant` if you're iterating over `VARIANTS`?